### PR TITLE
Added support extend the metadata information with the WebKit bugzilla

### PR DESCRIPTION
**Pull request by psaavedra**
Related to https://github.com/web-platform-tests/wpt-pr-bot/issues/100

* Added webkit metadata library
* Added bugs-webkit library for WebKit's Bugzilla
* Added pullRequestPuller() in the main index.js. This function pulls
  the "Webkit export" open PR and updates the metadata info verifiying
  the PR against the Bugzilla
* getMedatada() extended to include WebKit related items:
  * For WebKit's Bugzilla flags: in_queue and reviewed
  * For WebKit's Bugzilla see_also link reference

### CODE REVIEW

File: lib/bugs-webkit.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    body = JSON.parse(body);
+    var err = new Error(body.message);
+    if (body.errors) { err.errors = body.errors; }
+    err.statusCode = response.statusCode;
+    return err;
+}
+
+function getHeaders() {
+    return {
+        "User-Agent": "ganesh",
```
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply lib/bugs-webkit.js]: I see now that this is copied from github.js. Fine for now then.

File: index.js [hunk 2]
```diff
@@ -125,3 +128,57 @@ app.listen(port, function() {
     if (flags.get('dry-run'))
         console.log('Starting in DRY-RUN mode');
 });
+
+function pullRequestPuller() {
+    waitFor(5 * 1000).then(function() {
```
[the remark this answers is not in the document]
Comment (psaavedra) [reply index.js]: I don't have a strong opinion about but yeah maybe minutes is better than seconds having in account the purpose of this code.

File: index.js [hunk 3]
```diff
@@ -125,3 +128,57 @@ app.listen(port, function() {
...
+                metadata(pull_request.number, pull_request.user.login, pull_request.title, pull_request.body).then(function(metadata) {
+                    // logArgs(metadata);
+                    console.log("Issue: " + metadata.issue);
+                    console.log("Title: " + metadata.title);
+                    console.log("Labels: " + metadata.labels);
+                    console.log("isWebKitVerified: " + metadata.isWebKitVerified);
+                    console.log("isMergeable: " + metadata.isMergeable);
+                    console.log("reviewedDownstream: " + metadata.reviewedDownstream);
+
+                    if (metadata.isWebKitVerified) {
```
[the remark this answers is not in the document]
Comment (psaavedra) [reply index.js]: Yes, these flags provides the current status of a patch in the WebKit side.

File: test/get-metadata.js [hunk 4]
```diff
@@ -366,6 +370,94 @@ suite('getMetadata', function() {
...
+            isWebKitVerified: false,
+            isMergeable: true,
+            reviewedDownstream: null,
+            missingReviewers: [],
+            missingAssignee: null ,
+        };
+
+        return getMetadata(19538, 'ntrrgc', 'WebKit export of https://bugs.webkit.org/show_bug.cgi?id=201401', '')
+            .then(function(actual) {
+                console.log(actual);
```
[the remark this answers is not in the document]
Comment (psaavedra) [reply test/get-metadata.js]: sure it was there by mistake

**Comment by psaavedra** [comment]
I was bit stuck whit the tests but I think I got it now.

File: index.js [hunk 5]
```diff
@@ -125,3 +128,57 @@ app.listen(port, function() {
...
+                         funkLogErr(n, "Something went wrong while adding missing LABELS.")
+                    ).then(function() {
+                        return comment(n, metadata);
+                    }).then(
+                        funkLogMsg(n, "Added missing REVIEWERS if any."),
+                        funkLogErr(n, "Something went wrong while adding missing REVIEWERS.")
+                    ); */
+                });
+			});
+            pullRequestPuller();
```
[the remark this answers is not in the document]
Comment (psaavedra) [reply index.js]: I think it is safe. the pullRequestPuller()  relays in the an async_func/then/promise chain. 
The pullRequestPuller() functions simply doesn't wait for the async functions so reach its end and just the promises is added the callback chains because the then() hadlers:

   waitFor(60 * 1000).then() -> github.get().then()

Just when the callback is resolved the recursive pullRequestPuller() call is invoked but the previous one was gone time ago.

[the remark this answers is not in the document]
Comment (psaavedra) [reply index.js]: in the top of that,  the memory. I also ran a profiling with the `npm --inspect` for 15 minutes in order to check memory allocations leaks (heap and stack). The graph remains looks good:

Screenshot from 2019-12-19 14-13-28

File: package.json [hunk 6]
```diff
@@ -19,6 +19,7 @@
   "dependencies": {
     "@google-cloud/datastore": "^5.0.1",
     "bl": "^4.0.0",
+    "core-js": "^2.6.10",
```
[the remark this answers is not in the document]
Comment (psaavedra) [reply package.json]: yeah, It seems I added it for no reason. I suppose I did it in the beginning I just I did realize about it when I commit the change. I just double check the installation from the scratch and all works as we expected without this depends.

File: lib/bugs-webkit.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
+"use strict";
+var token = process.env.WEBKIT_BUGZILLA_TOKEN;
```
[the remark this answers is not in the document]
Comment (psaavedra) [reply lib/bugs-webkit.js]: 2be99b283a89323a9d615d3628238ab77da2d8d5 , right?

[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply lib/bugs-webkit.js]: Correct. This looks good to me; we can sync on the token value and make sure it is stored in the cloud datastore before we deploy this to production.

File: index.js [hunk 8]
```diff
@@ -125,3 +131,49 @@ app.listen(port, function() {
...
+// set of PRs to keep WebKit exports synchronized with the upstream PR.
+function pullRequestPoller() {
+    waitFor(60 * 1000).then(function() {
+        github.get("/repos/:owner/:repo/pulls", {}).then(function (pull_requests) {
+            pull_requests.forEach(function(pull_request) {
+                if (!webkit.related(pull_request.title)) {
+                    return;
+                }
+                metadata(pull_request.number, pull_request.user.login, pull_request.title, pull_request.body).then(function(metadata) {
+                    // logArgs(metadata);
```
Comment (stephenmcgruer) [inline index.js]: Nit; remove unused line.

File: index.js [hunk 9]
```diff
@@ -125,3 +131,49 @@ app.listen(port, function() {
...
+                        console.log("Webkit flags:");
+                        if (metadata.webkit.flags.reviewed) {
+                              console.log("  Reviewed");
+                        }
+                        if (metadata.webkit.flags.inCommit) {
+                              console.log("  inCommit");
+                        }
+                    }
+                    // TODO: This part of the code must be uncommmented to
+                    /* return labelModel.post(n, metadata.labels).then(
```
Comment (stephenmcgruer) [inline index.js]:
```suggestion
                    /* return labelModel.post(n, metadata.labels, flags.get('dry-run')).then(
```

File: index.js [hunk 10]
```diff
@@ -125,3 +131,49 @@ app.listen(port, function() {
...
+                         funkLogMsg(n, "Added missing LABELS if any."),
+                         funkLogErr(n, "Something went wrong while adding missing LABELS.")
+                    ).then(function() {
+                        return comment(n, metadata);
```
Comment (stephenmcgruer) [inline index.js]:
```suggestion
                        return comment(n, metadata, flags.get('dry-run'));
```

File: lib/filter.js [hunk 11]
```diff
@@ -36,5 +36,19 @@ function filterPullRequest(pull_request, log) {
     return true;
 }
 
+function filterPullRequestByLabel(pull_request, label, log) {
```
Comment (stephenmcgruer) [inline lib/filter.js]: This is now dead code I believe (since we aren't dealing with 'do not merge'); please remove it.

File: lib/metadata/webkit.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+};
+
+exports.flags = function(issue) {
+        var flags = {
+            reviewed: false,
+            inCommit: false
+        };
+        return bugsWebkit.get("/rest/bug/:id/history", { id: issue }).then(function (body) {
+            body.bugs[0].history.forEach(function(history_item) {
+                // console.log("When:" + history_item.when);
```
Comment (stephenmcgruer) [inline lib/metadata/webkit.js]: Nit; please remove unused console log lines (here and three lines just below)

File: lib/metadata/webkit.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+                history_item.changes.forEach(function(change) {
+                    // console.log("Change - added:" + change.added);
+                    // console.log("Change - removed:" + change.removed);
+                    // console.log("Change - field_name:" + change.field_name);
+                    if (change.field_name == "flagtypes.name") {
+                        if (change.removed.includes("review+")) {
+                            flags.reviewed = false;
+                        }
+                        if (change.removed.includes("commit-queue+")) {
+                            flags.reviewed = false;
```
Comment (stephenmcgruer) [inline lib/metadata/webkit.js]: Why does taking a WebKit bug off of the commit-queue imply it is no longer reviewed?

File: lib/metadata/webkit.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+                        if (change.added.includes("review+")) {
+                            flags.reviewed = true;
+                        }
+                        if (change.added.includes("commit-queue+")) {
+                            flags.reviewed = true;
+                            flags.inCommit = true;
+                        }
+                    }
+                    if (change.field_name == "resolution") {
+                        if (change.added.includes("FIXED")) {
```
Comment (stephenmcgruer) [inline lib/metadata/webkit.js]: Is it possible for a WebKit bug to be closed FIXED without the change actually being committed to the codebase?

File: lib/metadata/webkit.js [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+                            flags.reviewed = true;
+                            flags.inCommit = true;
+                        }
+                    }
+                });
+            });
+            // console.log("flags.reviewed: " + flags.reviewed);
```
Comment (stephenmcgruer) [inline lib/metadata/webkit.js]: Nit; please remove these two unused lines

File: test/filter.js [hunk 16]
```diff
@@ -64,3 +64,23 @@ suite('Pull request filtering', function() {
         assert.strictEqual(filter.pullRequest({mergeable_state: "unknown"}, nullLog), true);
     });
 });
+
+suite('Pull request filtering by label', function() {
```
Comment (stephenmcgruer) [inline test/filter.js]: Since the code this tests is dead, please remove.

File: lib/metadata/index.js [hunk 17]
```diff
@@ -29,18 +30,16 @@ function inferDownstreamReview(metadata, content) {
...
-        login == "youennf")) {
+    if (metadata.isWebKitVerified && metadata.webkit.flags.reviewed) {
         return "WebKit";
     }
-
     return null;
 }
 
-module.exports = function getMetadada(number, author, content) {
+module.exports = function getMetadada(number, author, title, content) {
```
Comment (stephenmcgruer) [inline lib/metadata/index.js]: I'm not sure if all callsites to this have been updated. For example, in `test/get-metadata.js`, shouldn't every call to `getMetadata` have been updated? (As otherwise they are now passing in their previous content for the title, and `undefined` for the content).

Can you verify everywhere that pulls in `lib/metadata` and uses it is passing the right arguments?

File: test/get-metadata.js [hunk 18]
```diff
@@ -366,6 +370,101 @@ suite('getMetadata', function() {
...
+        return getMetadata(19538, 'ntrrgc', 'WebKit export of https://bugs.webkit.org/show_bug.cgi?id=201401', '')
+            .then(function(actual) {
+                assert.deepEqual(expected, actual);
+            });
+    });
+    test('retrieval and formatting of metadata for a WebKit related pull request with wrong bugzilla reference', function() {
+        var expected = "'202311' is not a valid bug number nor an alias to a bug.";
+        return getMetadata(19381, 'rwlbuis', 'WebKit export of https://bugs.webkit.org/show_bug.cgi?id=202311', '')
+            .then(function() {}, function(actual) {
+                console.log(actual);
```
Comment (stephenmcgruer) [inline test/get-metadata.js]: Another leftover log; please remove

File: lib/metadata/webkit.js (hunk 13 above)
Comment (psaavedra) [reply lib/metadata/webkit.js]: It is a corner case: _Commit removed from the commit queue before commit_ 


...  in other case (`cq+`  finally landed) ..., the `cq+`  flag is also removed but the issues is marked as *fixed* ( `change.added.includes("FIXED")`) so tit keeps the values in:

```
               flags.reviewed = true;
               flags.inCommit = true;
```

File: lib/metadata/webkit.js (hunk 14 above)
Comment (psaavedra) [reply lib/metadata/webkit.js]: Yes. It is possible set it as `Fixed` manually. 

I will change the condition as follows:

``` diff
@@ -45,7 +44,7 @@ exports.flags = function(issue) {
                         }
                     }
                     if (change.field_name == "resolution") {
-                        if (change.added.includes("FIXED")) {
+                        if (change.added.includes("FIXED") && change.who == "commit-queue@webkit.org" ) {
                             flags.reviewed = true;
                             flags.inCommit = true;
                         }
```

to ensure that the Fixed was done by the commit-queue.

File: test/get-metadata.js (hunk 18 above)
Comment (psaavedra) [reply test/get-metadata.js]: oops

Comment (psaavedra) [reply test/get-metadata.js]: fixed

**Comment by psaavedra** [comment]
> Sorry for the delay. I think we're getting there! :)

Nevermind. Thanks for the review. I just update the PR with the suggested changes. I feel we are pretty close.

File: lib/metadata/webkit.js [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+        var flags = {
+            reviewed: false,
+            inCommit: false
+        };
+        return bugsWebkit.get("/rest/bug/:id/history", { id: issue }).then(function (body) {
+            body.bugs[0].history.forEach(function(history_item) {
+                history_item.changes.forEach(function(change) {
+                    // console.log("Change - added:" + change.added);
+                    // console.log("Change - removed:" + change.removed);
+                    // console.log("Change - field_name:" + change.field_name);
```
Comment (stephenmcgruer) [inline lib/metadata/webkit.js]: Nit; please remove commented-out log lines.

File: index.js [hunk 20]
```diff
@@ -125,3 +131,48 @@ app.listen(port, function() {
...
+                    if (metadata.isWebKitVerified) {
+                        console.log("Webkit flags:");
+                        if (metadata.webkit.flags.reviewed) {
+                              console.log("  Reviewed");
+                        }
+                        if (metadata.webkit.flags.inCommit) {
+                              console.log("  inCommit");
+                        }
+                    }
+                    // TODO: This part of the code must be uncommmented to
```
Comment (stephenmcgruer) [inline index.js]: I'm will land the uncomment of this myself in a follow-up commit once I've validated the change locally.

Comment (psaavedra) [reply index.js]: OK. Let me know if I could do some additional verification in this regard
